### ISSUE TITLE
Should aggregates be grouped by fields that occur in conditions?

### ISSUE THREAD

**Issue by jakevdp** [vega/vega-lite#6045]
This is followup from an Altair user question that comes from a potentially confusing aspect of the grammar.

TLDR: In the VL grammar, aggregates specified in encodings are implicitly grouped by other encodings. Should they also be grouped by conditions that appear in those encodings?

Consider this chart ([vega editor](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAmCGAutIC4zEgVwE4BsVTkQHoBjWLAZwDoArCgewDtIBfAGnCgFtyBrfSACNykDhEgBTRiXrQAlowDm+UBHEyc9LCsgAzORJzQBAYQCeOBdAmVRUeGYAOEgVvmNYedp3EAPHfqGxqiQALJyOBIUAPrOWNEA4p6azGxQggr48FgYEmmQDs4CAI4YsIzwcoiVAG4u3mpQZjqwiopYEooILiEyGBV2BU49UKXlldVydaycLCAsQA)):
```json
{
  "data": {"url": "data/cars.json"},
  "mark": "bar",
  "encoding": {
    "x": {"field": "Miles_per_Gallon", "bin": true, "type": "quantitative"},
    "y": {"aggregate": "count", "type": "quantitative"},
    "color": {"field": "Cylinders", "type": "ordinal"}
  }
}
```
![visualization - 2020-03-09T114912 976](https://user-images.githubusercontent.com/781659/76246994-031b6000-61fc-11ea-9536-4886f6f487b9.png)

Now suppose the user wants to highlight rows with fewer than 5 cylinders. They might look at the docs and try replacing the color encoding with [condition](https://vega.github.io/vega-lite/docs/condition.html) ([editor](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAmCGAutIC4zEgVwE4BsVTkQHoBjWLAZwDoArCgewDtIBfAGnCgFtyBrfSACNykDhEgBTRiXrQAlowDm+UBHEAPFZABmciTmgCAsnJwSKAfQAOErBYDisHDiaioghfnhYMEtlHgATxsBAEcMWEZ4OURogDcJVjE1SECtWEVFLAlFBETUSBkMKLdIIJCC8Mjo2LkEpM5xGRcsFUaUmUZ5aNdUdDinXwEKeAl9QRwh-zLzeAFCDC4qAGFAnAVoWwowAB4wAFYGtRSByfyCPnosSMVE9pZOB5YgA)):
```json
{
  "data": {"url": "data/cars.json"},
  "mark": "bar",
  "encoding": {
    "x": {"field": "Miles_per_Gallon", "bin": true, "type": "quantitative"},
    "y": {"aggregate": "count", "type": "quantitative"},
    "color": {
      "condition": {"test": "datum.Cylinders < 5", "value": "steelblue"},
      "value": "darkorange"
    }
  }
}
```
![visualization - 2020-03-09T114858 035](https://user-images.githubusercontent.com/781659/76246999-057dba00-61fc-11ea-92fc-9a1b88cf9632.png)

This clearly does not have the desired effect, because the count aggregate is no longer grouped by ``Cylinders``. For users unfamiliar with the details of how aggregates are computed in VL, it's quite difficult to debug why this is happening.

One easy remedy is to explicitly add a ``detail`` encoding, so that the counts are appropriately grouped ([editor](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAmCGAutIC4zEgVwE4BsVTkQHoBjWLAZwDoArCgewDtIBfAGnCgFtyBrfSACNykDhEgBTRiXrQAlowDm+UBHEAPFZABmciTmgCAsnJwSKAfQAOErBYDisHDiaioghfnhYMEtlHgATxsBAEcMWEZ4OURogDcJVjE1SECtWEVFLAlFBETUSBkMKLdIIJCC8Mjo2LkEpM5xaAlEUy1dfUMCgGFAnAVmylLy-Kh6LHlGJwa1KBkXLBVGlJlGeWjXVHQ4p18BCngJfUEcPf8y83gBQgwuKl7+tdsKMAAeMABWGdmoHdPRmB8caRRSJZYsTgQlhAA)):
```json
{
  "data": {"url": "data/cars.json"},
  "mark": "bar",
  "encoding": {
    "x": {"field": "Miles_per_Gallon", "bin": true, "type": "quantitative"},
    "y": {"aggregate": "count", "type": "quantitative"},
    "detail": {"field": "Cylinders", "type": "ordinal"},
    "color": {
      "condition": {"test": "datum.Cylinders < 5", "value": "steelblue"},
      "value": "darkorange"
    }
  }
}
```
![visualization - 2020-03-09T115121 569](https://user-images.githubusercontent.com/781659/76247139-4d9cdc80-61fc-11ea-9c20-29db7e6bf4fd.png)

A more complete approach would probably be to apply a calculate transform and encode the color by that field ([editor](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAmCGAutIC4zEgVwE4BsVTkQHoBjWLAZwDoArCgewDtIBfAGnCni1kYoDN6WALb4A2pwjoyOEhhwIApvhgIMwqgGEAnjgCWjaIspgAPGACsYAPxgA5OYt2wqO4FMiS3chsosCip19Q2N-Fk4AXQ4ISGFyAGsVACNyb05IRUYSemgDAHN8UAhogA8CyH49RRxoFQBZPRxFCgB9AAdjZoBxWBwcJm8oRIN8bgxFH0h4bXaVAEcMXng9RCWAN2V2SShtMthc3KxFXKUVLIxGeAHJ6eVUSHnF5YQ9ddYoosgsvqwyiqqau6BAxGShXKYzO6MejCAw9VicMIsIA)):
```json
{
  "data": {"url": "data/cars.json"},
  "transform": [
    {"calculate": "datum.Cylinders < 5 ? '< 5' : '≥ 5'", "as": "Cylinders"}
  ],
  "mark": "bar",
  "encoding": {
    "x": {"field": "Miles_per_Gallon", "bin": true, "type": "quantitative"},
    "y": {"aggregate": "count", "type": "quantitative"},
    "color": {"field": "Cylinders", "type": "nominal"}
  }
}
```
![visualization - 2020-03-09T115540 120](https://user-images.githubusercontent.com/781659/76247536-f1868800-61fc-11ea-8576-0103acd01714.png)

But this is probably more suited to a polished, final chart than to quick and dirty data exploration.

Would it make sense to change the grammar such that aggregates specified in encodings will also group by fields that appear in conditional statements? In other words, should we treat fields referenced in conditional expressions as if they are included in the detail encoding?

**Comment by kanitw** [comment]
It's probably reasonable to include a field included in the test in the groupby.  

If anything I think we should make the following spec:
```
{
  "data": {"url": "data/cars.json"},
  "mark": "bar",
  "encoding": {
    "x": {"field": "Miles_per_Gallon", "bin": true, "type": "quantitative"},
    "y": {"aggregate": "count", "type": "quantitative"},
    "color": {
      "condition": {"test": {"field": "Cylinders", "lt": 5}, "value": "steelblue"},
      "value": "darkorange"
    }
  }
}
```

However, implementing such parsing for expression would make the code perhaps too complicated.
